Middleware  NextRequest params type error

### What is the improvement or update you wish to see?

https://next-auth.js.org/configuration/nextjs#wrap-middleware

In this documentation, wrap middleware get `req: NextRequest & { nextauth: { token: JWT | null } }` but this code got typescript error.

```
error TS2345: Argument of type '[(req: NextRequest & { nextauth: { token: JWT; }; }) => void]' is not assignable to parameter of type 'WithAuthArgs'.
  Type '[(req: NextRequest & { nextauth: { token: JWT; }; }) => void]' is not assignable to type '[NextRequest]'.
    Type '(req: NextRequest & { nextauth: { token: JWT; }; }) => void' is not assignable to type 'NextRequest'.
```

### Is there any context that might help us understand?

I think this issue has some options.

* Improve code types: maybe like `NextMiddlewareWithJWT`
* Improve doc: append `@ts-ignore` ( for workaround )

### Does the docs page already exist? Please link to it.

https://next-auth.js.org/configuration/nextjs#wrap-middleware

--- a/apps/dev/middleware.ts
+++ b/apps/dev/middleware.ts
@@ -30,12 +30,11 @@ export const config = { matcher: ["/middleware-protected"] }
 // export default withAuth(
 //   function middleware(req, ev) {
 //     console.log(req, ev)
-//     return undefined // NOTE: `NextMiddleware` should allow returning `void`
 //   },
 //   {
 //     callbacks: {
 //       authorized: ({ token }) => token.name === "Balázs Orbán",
-//     }
+//     },
 //   }
 // )
 

--- a/docs/docs/configuration/nextjs.md
+++ b/docs/docs/configuration/nextjs.md
@@ -177,13 +177,11 @@ If you do not define the options, NextAuth.js will use the default values for th
 #### wrap middleware
 
 ```ts title="middleware.ts"
-import type { NextRequest } from "next/server"
-import type { JWT } from "next-auth/jwt"
 import { withAuth } from "next-auth/middleware"
 
 export default withAuth(
-  // `withAuth` can augment your Request with the user's token.
-  function middleware(req: NextRequest & { nextauth: { token: JWT | null } }) {
+  // `withAuth` augments your `Request` with the user's token.
+  function middleware(req) {
     console.log(req.nextauth.token)
   },
   {

--- a/packages/next-auth/src/next/middleware.ts
+++ b/packages/next-auth/src/next/middleware.ts
@@ -189,12 +189,10 @@ export function withAuth(...args: WithAuthArgs) {
   if (typeof args[0] === "function") {
     const middleware = args[0]
     const options = args[1] as NextAuthMiddlewareOptions | undefined
-    return async (...args: Parameters<NextMiddleware>) =>
+    return async (...args: Parameters<NextMiddlewareWithAuth>) =>
       await handleMiddleware(args[0], options, async (token) => {
-        const [req, ...rest] = args
-        // @ts-expect-error
-        req.nextauth = { token }
-        return await middleware(req as NextRequestWithAuth, ...rest)
+        args[0].nextauth = { token }
+        return await middleware(...args)
       })
   }
 

--- a/packages/next-auth/src/next/middleware.ts
+++ b/packages/next-auth/src/next/middleware.ts
@@ -92,7 +92,9 @@ export interface NextAuthMiddlewareOptions {
   secret?: string
 }
 
-type NextMiddlewareResult = ReturnType<NextMiddleware>
+// TODO: `NextMiddleware` should allow returning `void`
+// Simplify when https://github.com/vercel/next.js/pull/38625 is merged.
+type NextMiddlewareResult = ReturnType<NextMiddleware> | void
 
 async function handleMiddleware(
   req: NextRequest,
@@ -145,12 +147,21 @@ async function handleMiddleware(
   return NextResponse.redirect(signInUrl)
 }
 
+export interface NextRequestWithAuth extends NextRequest {
+  nextauth: { token: JWT | null }
+}
+
+export type NextMiddlewareWithAuth = (
+  request: NextRequestWithAuth,
+  event: NextFetchEvent
+) => NextMiddlewareResult | Promise<NextMiddlewareResult>
+
 export type WithAuthArgs =
-  | [NextRequest]
-  | [NextRequest, NextFetchEvent]
-  | [NextRequest, NextAuthMiddlewareOptions]
-  | [NextMiddleware]
-  | [NextMiddleware, NextAuthMiddlewareOptions]
+  | [NextRequestWithAuth]
+  | [NextRequestWithAuth, NextFetchEvent]
+  | [NextRequestWithAuth, NextAuthMiddlewareOptions]
+  | [NextMiddlewareWithAuth]
+  | [NextMiddlewareWithAuth, NextAuthMiddlewareOptions]
   | [NextAuthMiddlewareOptions]
   | []
 
@@ -180,8 +191,10 @@ export function withAuth(...args: WithAuthArgs) {
     const options = args[1] as NextAuthMiddlewareOptions | undefined
     return async (...args: Parameters<NextMiddleware>) =>
       await handleMiddleware(args[0], options, async (token) => {
-        ;(args[0] as any).nextauth = { token }
-        return await middleware(...args)
+        const [req, ...rest] = args
+        // @ts-expect-error
+        req.nextauth = { token }
+        return await middleware(req as NextRequestWithAuth, ...rest)
       })
   }
 

--- a/packages/next-auth/src/next/middleware.ts
+++ b/packages/next-auth/src/next/middleware.ts
@@ -94,7 +94,7 @@ export interface NextAuthMiddlewareOptions {
 
 // TODO: `NextMiddleware` should allow returning `void`
 // Simplify when https://github.com/vercel/next.js/pull/38625 is merged.
-type NextMiddlewareResult = ReturnType<NextMiddleware> | void
+type NextMiddlewareResult = ReturnType<NextMiddleware> | void // eslint-disable-line @typescript-eslint/no-invalid-void-type
 
 async function handleMiddleware(
   req: NextRequest,

